src/scripts/Invoke-RegisterSubscriptionResourceProviders.ps1

```diff
@@ -0,0 +1,84 @@
+param(
+  [string]$subscriptionId,
+  [string]$resourceProviders
+)
+
+$ErrorActionPreference = "SilentlyContinue"
+# Selecting the right subscription
+Select-AzSubscription -SubscriptionId $subscriptionId
+
+# Defining variables
+$providers = $resourceProviders | ConvertFrom-Json -AsHashtable
+$failedProviders = ""
+$failedFeatures = ""
+$DeploymentScriptOutputs = @{}
+
+##############################################
+## Registering resource providers and features
+##############################################
+
+if ($providers.Count -gt 0) {
+  foreach ($provider in $providers.keys) {
+    try {
+      # Registering resource providers
+      $providerStatus = (Get-AzResourceProvider -ListAvailable | Where-Object ProviderNamespace -eq $provider).registrationState
+      # Check if the providered is registered
+      if ($providerStatus -ne 'Registered') {
+        Write-Output "`n Registering the '$provider' provider"
+        if (Register-AzResourceProvider -ProviderNamespace $provider) {
```

Comment:
This code might introduce some issues that will cause problems later:

`Register-AzResourceProvider` is not synchronous operation with the actual registration process.`Register-AzResourceProvider` returns as soon as management API accepts HTTP POST request. Worst case scenario caused by this is failures with next Bicep deployments that are using resource providers that are still being on `Registering` `RegistrationState`

Issue number two is variation of previous. Code does not check if registration process ends with `registrationState`  = `Registered`

----
If those issues are fixed by writing polling code with `Get-AzResourceProvider`
 or running `Register-AzResourceProvider` with `AsJob` then the next step would be adding parallelism into code. Otherwise full registration process will be really slow.
 ----
Same issue is also present with `Register-AzProviderFeature`

Reply:
Thanks for sharing this feedback. We have some logic to output the providers that have not been registered successfully so the user can take action. We decided to only cater for ARM errors when starting the registration providers and not to wait because it can take extended amounts of time. We have added some logic based on your feedback to include providers in "registering" state to the output variable as well

Reply:
It might be a good idea to document that  `src/self/subResourceWrapper/deploy.bicep`:`registerResourceProviders` is not fully synchronous. Just to make sure that module users does not make an assumption that eveything is up and running after main.bicep. deployment is completed.